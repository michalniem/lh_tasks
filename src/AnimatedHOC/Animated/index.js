import React, { useRef } from "react";

import "./style.css";
import getAnimationPropeprty from "./helpers/getAnimationProperty";
import useVisibilitySensor from "./hooks/useVisibilitySensor";

export function Animated({
  children,
  options = {
    animation: {
      name: "fade",
      duration: "1s",
      delay: "0.2s",
      timingFunction: "ease-in-out",
    },
  },
}) {
  const containerRef = useRef(null);
  const isIntersecting = useVisibilitySensor(containerRef);

  return (
    <div
      ref={containerRef}
      style={{
        opacity: 0,
        animation: isIntersecting
          ? getAnimationPropeprty(options.animation)
          : "none",
      }}
    >
      {children}
    </div>
  );
}

export function WithAnimation(Component, options, propsToPass) {
  return class extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        isIntersecting: false,
      };
      this.containerRef = React.createRef();
      this.observer = new IntersectionObserver(([entry]) => {
        this.setState({ isIntersecting: entry.isIntersecting });
      }, options);
    }

    componentDidMount() {
      if (this.containerRef.current)
        this.observer.observe(this.containerRef.current);
    }

    componentWillUnmount() {
      if (this.containerRef.current)
        this.observer.unobserve(this.containerRef.current);
    }

    render() {
      return (
        <div ref={this.containerRef}>
          <Component {...propsToPass} isVisible={this.state.isVisible}>
            {this.props.children}
          </Component>
        </div>
      );
    }
  };
}

export default Animated;
