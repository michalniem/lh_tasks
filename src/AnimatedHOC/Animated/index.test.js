import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import Animated from "./index";

const NoopComponent = () => <div />;

describe("Animated", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Animated>
        <NoopComponent />
      </Animated>
    );
  });

  test("should render correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
