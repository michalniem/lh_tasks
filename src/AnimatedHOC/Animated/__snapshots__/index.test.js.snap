// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animated should render correctly 1`] = `
<div>
  <ForwardRef(MotionComponent)
    animate="exit"
    transition={
      Object {
        "duration": 0.3,
      }
    }
    variants={
      Object {
        "enter": Object {
          "opacity": 1,
        },
        "exit": Object {
          "opacity": 0,
        },
      }
    }
  >
    <NoopComponent />
  </ForwardRef(MotionComponent)>
</div>
`;
